Block in dlopen until all threads have loaded the module (#18376)
emscripten-core/emscripten: site/source/docs/compiling/Dynamic-Linking.rst

Pthreads support
----------------

Dynamic linking + pthreads is is still experimental.  As such, linking with both
``MAIN_MODULE`` and ``-pthread`` will produce a warning.

While load-time dynamic linking works without any complications, runtime dynamic
linking via ``dlopen``/``dlsym`` can require some extra consideration.  The
reason for this is that keeping the indirection function pointer table in sync

> between threads has to be done my emscripten library code.  Each time a new

Reviewer: ```suggestion
between threads has to be done by emscripten library code.  Each time a new
```

> between threads has to be done my emscripten library code.  Each time a new

Reviewer 2: ```suggestion
between threads has to be done by emscripten library code.  Each time a new
```

> library is loaded of a new symbol is requested via ``dlsym``, table slots can be

Reviewer: ```suggestion
library is loaded or a new symbol is requested via ``dlsym``, table slots can be
```

> library is loaded of a new symbol is requested via ``dlsym``, table slots can be

Reviewer 2: ```suggestion
library is loaded or a new symbol is requested via ``dlsym``, table slots can be
```

added and these changes need to be mirrored on every thread in the process.

Changes to the table are protected by a mutex, and before any thread returns
from ``dlopen`` or ``dlsym`` it will wait until all other threads are sync.  In

> order to make this synchronization as seamless as possible we hook into the

Reviewer: ```suggestion
order to make this synchronization as seamless as possible, we hook into the
```

> low level primitives of `emscripten_futex_wait` and `emscirpten_yeild`.

Reviewer: ```suggestion
low level primitives of `emscripten_futex_wait` and `emscripten_yield`.
```

> low level primitives of `emscripten_futex_wait` and `emscirpten_yeild`.

Reviewer 2: ```suggestion
low level primitives of `emscripten_futex_wait` and `emscripten_yield`.
```

For most use cases all this happens under hood and no special action is needed.

> However, there there is one classes of application that currently may require

Reviewer: ```suggestion
However, there there is one class of application that currently may require
```

> modification.  If your applications busy waits, or directly use the

Reviewer: ```suggestion
modification.  If your applications busy waits, or directly uses the
```

> modification.  If your applications busy waits, or directly use the

Reviewer 2: ```suggestion
modification.  If your applications busy waits, or directly uses the
```

``atomic.waitXX`` instructions (or the clang
``__builtin_wasm_memory_atomic_waitXX`` builtins) you maybe need to switch it

> using ``emscripten_futex_wait`` to avoid deadlocks.  If you don't use

Reviewer: ```suggestion
to use ``emscripten_futex_wait`` to avoid deadlocks.  If you don't use
```

> using ``emscripten_futex_wait`` to avoid deadlocks.  If you don't use

Reviewer 2: ```suggestion
to use ``emscripten_futex_wait`` to avoid deadlocks.  If you don't use
```
